## Final Project Submission

Please fill out:
* Student name: COLLINS KANYIRI MWANGI
* Student pace: full time
* Scheduled project review date/time: 
* Instructor name: Veronica Isiaho
* Blog post URL:


# Multiple Linear Regression: Predicting King County, WA, Housing Prices

## 1.0 BUSINESS UNDERSTANDING

### 1.1 Understanding the Problem

The role assumed here is of a Data Scientist working for Neocivil Realtors. The Agency is in the business of helping homeowners buy and/or sell houses. The Agency and the homeowners reside in King County in the US State of Washington. The Agency has provided me with a dataset of house sales in King County. The task here is to model the real estate housing prices and use the model to accurately predict the housing prices based on a number of features provided within the dataset. The expectation is that, once completed, the model can be used by them as a tool in selecting properties for investment in King County.

## 1.2 Analysis Questions

This analysis will seek to answer three questions about the data:

     Question 1: Which features are most highly correlated with price?

     Question 2: Which features have the strongest correlations with other predictor variables?

     Question 3: What combinations of features is the best fit, in terms of predictive power, for a multiple regression model to                  predict house prices based on provided historical data?
     

## 2.0 DATA UNDERSTANDING

id - unique identified for a house

date - Date house was sold

price - Price is prediction target

bedroomsNumber - of Bedrooms/House

bathroomsNumber - of bathrooms/bedrooms

sqft_livingsquare - footage of the home

sqft_lotsquare - footage of the lot

floorsTotal - floors (levels) in house

waterfront - House which has a view to a waterfront

view - Has been viewed

condition - How good the condition is ( Overall )

grade - overall grade given to the housing unit, based on King County grading system

sqft_above - square footage of house apart from basement

sqft_basement - square footage of the basement

yr_built - Built Year

yr_renovated - Year when house was renovated

zipcode - zip

lat - Latitude coordinate

long - Longitude coordinate

sqft_living15 - The square footage of interior housing living space for the nearest 15 neighbors

sqft_lot15 - The square footage of the land lots of the nearest 15 neighbors

## 3.0 DATA PREPARATION

### 3.1 Importing Libraries

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
from sklearn.metrics import mean_absolute_error
import statsmodels.api as sm
from statsmodels.stats.outliers_influence import variance_inflation_factor
from statsmodels.formula.api import ols
from sklearn.model_selection import train_test_split
import warnings
warnings.filterwarnings('ignore')


### 3.2 Load and Exploring the data

In [2]:
# load the data
df = pd.read_csv('data/kc_house_data.csv')
# Preview of the first and last rows of the dataset
df

id        date     price  bedrooms  bathrooms  sqft_living   
0      7129300520  10/13/2014  221900.0         3       1.00         1180  \
1      6414100192   12/9/2014  538000.0         3       2.25         2570   
2      5631500400   2/25/2015  180000.0         2       1.00          770   
3      2487200875   12/9/2014  604000.0         4       3.00         1960   
4      1954400510   2/18/2015  510000.0         3       2.00         1680   
...           ...         ...       ...       ...        ...          ...   
21592   263000018   5/21/2014  360000.0         3       2.50         1530   
21593  6600060120   2/23/2015  400000.0         4       2.50         2310   
21594  1523300141   6/23/2014  402101.0         2       0.75         1020   
21595   291310100   1/16/2015  400000.0         3       2.50         1600   
21596  1523300157  10/15/2014  325000.0         2       0.75         1020   

       sqft_lot  floors waterfront  view  ...          grade sqft_above   
0          5650     1.0        NaN  NONE  ...      7 Average       1180  \
1          7242     2.0         NO  NONE  ...      7 Average       2170   
2         10000     1.0         NO  NONE  ...  6 Low Average        770   
3          5000     1.0         NO  NONE  ...      7 Average       1050   
4          8080     1.0         NO  NONE  ...         8 Good       1680   
...         ...     ...        ...   ...  ...            ...        ...   
21592      1131     3.0         NO  NONE  ...         8 Good       1530   
21593      5813     2.0         NO  NONE  ...         8 Good       2310   
21594      1350     2.0         NO  NONE  ...      7 Average       1020   
21595      2388     2.0        NaN  NONE  ...         8 Good       1600   
21596      1076     2.0         NO  NONE  ...      7 Average       1020   

       sqft_basement yr_built  yr_renovated  zipcode      lat     long   
0                0.0     1955           0.0    98178  47.5112 -122.257  \
1              400.0     1951        1991.0    98125  47.7210 -122.319   
2                0.0     1933           NaN    98028  47.7379 -122.233   
3              910.0     1965           0.0    98136  47.5208 -122.393   
4                0.0     1987           0.0    98074  47.6168 -122.045   
...              ...      ...           ...      ...      ...      ...   
21592            0.0     2009           0.0    98103  47.6993 -122.346   
21593            0.0     2014           0.0    98146  47.5107 -122.362   
21594            0.0     2009           0.0    98144  47.5944 -122.299   
21595            0.0     2004           0.0    98027  47.5345 -122.069   
21596            0.0     2008           0.0    98144  47.5941 -122.299   

       sqft_living15  sqft_lot15  
0               1340        5650  
1               1690        7639  
2               2720        8062  
3               1360        5000  
4               1800        7503  
...              ...         ...  
21592           1530        1509  
21593           1830        7200  
21594           1020        2007  
21595           1410        1287  
21596           1020        1357  

[21597 rows x 21 columns]

In [3]:
# Checked to see the datatype, number of columns and rows of the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21597 entries, 0 to 21596
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21597 non-null  int64  
 1   date           21597 non-null  object 
 2   price          21597 non-null  float64
 3   bedrooms       21597 non-null  int64  
 4   bathrooms      21597 non-null  float64
 5   sqft_living    21597 non-null  int64  
 6   sqft_lot       21597 non-null  int64  
 7   floors         21597 non-null  float64
 8   waterfront     19221 non-null  object 
 9   view           21534 non-null  object 
 10  condition      21597 non-null  object 
 11  grade          21597 non-null  object 
 12  sqft_above     21597 non-null  int64  
 13  sqft_basement  21597 non-null  object 
 14  yr_built       21597 non-null  int64  
 15  yr_renovated   17755 non-null  float64
 16  zipcode        21597 non-null  int64  
 17  lat            21597 non-null  float64
 18  long  

Most of the data types are numeric though we have a few object data types. The total number of rows is 21,597 and the number of columns in 21

In [4]:
# Checked the number of unique values for each column
df.nunique()

id               21420
date               372
price             3622
bedrooms            12
bathrooms           29
sqft_living       1034
sqft_lot          9776
floors               6
waterfront           2
view                 5
condition            5
grade               11
sqft_above         942
sqft_basement      304
yr_built           116
yr_renovated        70
zipcode             70
lat               5033
long               751
sqft_living15      777
sqft_lot15        8682
dtype: int64

There seems to be some repeated values in the id column.

In [ ]:
# perform a statistical summary of the data
df.describe()

Count: The count of non-null values in each column. It gives you an idea of missing values or potential data quality issues.

Mean: The average value of each column. It provides a measure of central tendency and can give you a sense of the typical value.

Standard Deviation: The measure of the spread or variability of each column's values around the mean. It indicates how dispersed the data points are.

Minimum and Maximum: The smallest and largest values in each column. It gives you the range of the data and helps identify potential outliers.

Quartiles (25%, 50%, and 75%): These values divide the data into four equal parts. The 50th percentile (median) represents the middle value, while the 25th and 75th percentiles indicate the lower and upper quartiles, respectively. They provide insights into the data's distribution and skewness.

In [ ]:
# check the toal number of null values in each column
df.isnull().sum()

The columns waterfront, view and yr_renovated have 2376, 63 and 3842 null values respectively.

## 4.0 DATA CLEANING

### 4.1 Missing Values

In [ ]:
# A look at the percentage of missing values
for col in df.columns:
    if df[col].isnull().sum() > 0:
        perc = (df[col].isnull().sum()/len(df[col]))*100
        print("The column", col,"has",df[col].isnull().sum(),"missing values, which is", round(perc, 1),"% of it's total")

#### Waterfront Column

In [ ]:
print(df['waterfront'].unique(),"\n")

The unique values are nan, NO and YES.

In [ ]:
print(df['waterfront'].value_counts(),"\n")

The most common value in this column is NO with 19075 entries while YES has only 146. This means most of these houses don't have a waterfront hence it's safe to assume that the ones with missing values also don't have a waterfront. Therefore, I will replace the missing values with NO.

In [ ]:
# replace missing values with 'NO'
df['waterfront'].fillna('NO',inplace=True)

# confirm if the missing values have been replaced
df['waterfront'].value_counts()

The NO entries have increased from 19075 to 21451  

In [ ]:
# check data type 
df['waterfront'].unique()

We can see that waterfront is of type object. This is not compatible for our model hence I need to convert it to type int. Before changing the datatype, I first need to change the column to a binary column with 1 == YES and 0 == NO

In [ ]:

# replace the values 'NO' and 'YES' with '0' and '1' respectively
df['waterfront'].replace({'NO': 0, 'YES': 1}, inplace=True)

#convert column to data type 'int'
df['waterfront'].astype(int)

#confirm if change occured
df['waterfront'].unique()

The values have now changed to 1s and 0s and the datatype is now integer

#### View Column

In [ ]:
print(df['view'].unique(),"\n")

The unique values:
'NONE' 'GOOD' 'EXCELLENT' 'AVERAGE' 'FAIR'
nan -- missing values

In [ ]:
print(df['view'].value_counts(),"\n")

The most common value in this column is NONE. So this means most of the houses don't have a view. It's safe to assume the 63 missing values also don't have a view. Therefore, I will replace the missing values with NONE

In [ ]:
print("Number of missing values:",df['view'].isnull().sum())

In [ ]:
# replace missing values with 'NONE'
df['view'].fillna('NONE',inplace=True)

# confirm if the missing values have been replaced
df['view'].value_counts()

The number of 'NONE" entries has increased from 19422 to 19485 thus showing that the missing 63 vales have been replaced by NONE.

#### Year Renovated Column

In [ ]:
print(df['yr_renovated'].unique(),"\n")

nan also exists in the unique values of this column. The years run from 1948 to 2014.

In [ ]:
print(df['yr_renovated'].value_counts(),"\n")

Number of distinct elements is: 70 
The most common value in this column is 0.0.
it's not possible to know what 0.0 stands for.

In [ ]:
# replace missing values with '0.0'
df['yr_renovated'].fillna(0.0,inplace=True)

# confirm if the missing values have been replaced
df['yr_renovated'].value_counts()

The number of missing values increased from 17011 to 20853

In [ ]:
# counter checking to see if there are any more missing values
df.isnull().sum()

No more missing values in the data

### 4.2 Conversion

#### Conversion of Date Column

In [ ]:
# split the date into month, day and year
date = df['date'].str.split('/', expand=True)

# create new columns for month and year and convert to integer
df['month_sold'] = date[0].astype(int)
df['yr_sold'] = date[2].astype(int)

# drop original date column
df.drop(columns=['date'], axis=1, inplace=True)

# check to see if changed were made
df.head()

#### Convert view to numerical

In [ ]:
df.view.value_counts()

In [ ]:
# make view feature 0, 1, 2, 3 or 4 instead of none, fair, average, good, excellent

df.view.replace({'NONE': 0, 'FAIR': 1, 'AVERAGE': 2, 'GOOD': 3, 'EXCELLENT': 4}, inplace=True)

In [ ]:
df.view.value_counts()

Replaced
'NONE' with 0, 'FAIR' with 1, 'AVERAGE' with 2, 'GOOD' with 3, 'EXCELLENT' with 4

#### Convert condition to numerical

In [ ]:
df.condition.value_counts()

In [ ]:
# make condition feature 0, 1, 2, 3 or 4 instead of poor, fair, average, good, very good

df.condition.replace({'Poor': 0, 'Fair': 1, 'Average': 2, 'Good': 3, 'Very Good': 4}, inplace=True)

In [ ]:
df.condition.value_counts()

Replaced 
'Poor' with 0, 'Fair' with 1, 'Average' with 2, 'Good' with 3, 'Very Good' with 4

#### Convert grade to numerical

In [ ]:
# replace grade description with the numerical grade at the beginning of the string, as an int

df['grade'] = df.grade.map(lambda x: x.split(' ')[0]).astype(float)

In [ ]:
df.grade.value_counts()

 Replace grade description with the numerical grade at the beginning of the string

#### Conversion of sqft_basement to Float/Int Type

In [ ]:
#5. Change column sqft_basement datatype into float

df['sqft_basement'] = [float(x) if x != '?' else 0.0 for x in df['sqft_basement']]
df.head()

we replaced the ? with 0.0, and converted it to float type

In [ ]:
#confirm if change occured
df['sqft_basement'].dtype

### Identify and deal with Outliers.

In [ ]:
# explore the number of unique characters in each column.
df.nunique(axis = 0)

In [ ]:
#plot the columns with more than single digits count.
fig, axs = plt.subplots(2,4, figsize = (15,6))
plt1 = sns.boxplot(df['price'], ax = axs[0,0])
plt2 = sns.boxplot(df['bedrooms'], ax = axs[0,1])
plt3 = sns.boxplot(df['bathrooms'], ax = axs[0,2])
plt4 = sns.boxplot(df['sqft_living'], ax = axs[0,3])
plt5 = sns.boxplot(df['sqft_lot'], ax = axs[1,0])
plt1 = sns.boxplot(df['floors'], ax = axs[1,1])
plt2 = sns.boxplot(df['sqft_above'], ax = axs[1,2])
plt3 = sns.boxplot(df['sqft_basement'], ax = axs[1,3])
plt.savefig('./images/fig1.png')

The above figures show that there are multipal columns contain some outlier data.

In [ ]:
to_modify = ['price', 'bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot','sqft_above','sqft_basement']
for col in to_modify:
    Q1 = df[col].quantile(0.25)
    Q3 = df[col].quantile(0.75)
    IQR = Q3 - Q1
    df = df[(df[col] >= Q1 - 1.5*IQR) & (df[col] <= Q3 + 1.5*IQR)]

In [ ]:
#Plot them again to check the results
# check the data after modification
fig, axs = plt.subplots(2,4, figsize = (15,6))
plt1 = sns.boxplot(df['price'], ax = axs[0,0])
plt2 = sns.boxplot(df['bedrooms'], ax = axs[0,1])
plt3 = sns.boxplot(df['bathrooms'], ax = axs[0,2])
plt4 = sns.boxplot(df['sqft_living'], ax = axs[0,3])
plt5 = sns.boxplot(df['sqft_lot'], ax = axs[1,0])
plt1 = sns.boxplot(df['floors'], ax = axs[1,1])
plt2 = sns.boxplot(df['sqft_above'], ax = axs[1,2])
plt3 = sns.boxplot(df['sqft_basement'], ax = axs[1,3])
plt.savefig('./images/fig2.png')

In [ ]:
#confirm if change occured
df.nunique(axis = 0)

The outlier values were removed from the data.

In [ ]:
# saving the cleaned data into a csv file
df.to_csv('data/cleaned_kchousing.csv', index=False)

### 5.0 STATISTICAL MODELING

#### 5.1 Checking Correlations

In [ ]:
#View heatmap to get  Variable Correlations
corr = df.corr().abs()
fig, ax=plt.subplots(figsize=(17,12))
fig.suptitle('Variable Correlations', fontsize=30, y=.95, fontname='DejaVu Sans')
heatmap = sns.heatmap(corr, cmap="YlGnBu", annot=True)
plt.savefig('images/Variable Correlations heatmap');
heatmap;

A correlation of 0.7 to 1.0 (-0.7 to -1) suggests a strong positive (negative) linear relationship while 0.5 to 0.6 (-0.5 to -0.6) is considered moderate.
many of the variables related to the size of homes (e.g. sqft_living, sqft_living15, etc) are strongly correlated with each other.

The target variable is price. Therefore, we look at the correlation coefficients for all of the predictor variables to find the one with the highest correlation with price.

In [ ]:
# correlation of the independent variables in descending order
df.corr()['price'].sort_values(ascending=False)

#### Question 1: Which features are most highly correlated with price?

sqft_living has the strongest positive correlation with price followed by 
 grade, sqft_living15, sqft_above and bathrooms in that order.
 NOTE. we do not use lat as its not needed in the analyisis
 

5.3.2 Plotting the sqft_living vs. Price

In [ ]:
# create a scatter plot of sqft_living vs. price:
fig, ax = plt.subplots(figsize=(8,6))
df.plot.scatter(x='sqft_living', y='price', ax=ax)
plt.title('Scatter Plot of Square Footage of Living Space vs Price');
plt.savefig('images/fig3');


### 5.2 Simple Linear Regression Model

####  Setting Up Variables for Regression

Declaring y and X_base variables, where y is a Series containing price data and X_base is a DataFrame containing the column with the target variable.

In [ ]:
y = df['price']

X_base = df['sqft_living']

#### Creating and Fitting Simple Linear Regression

In [ ]:
base_model = sm.OLS(endog=y, exog=sm.add_constant(X_base))
base_results = base_model.fit()

In [ ]:
#### Evaluate and Interpret Baseline Model Results
print(base_results.summary())

In [ ]:
# calculate the mean absolute error
base_mae = mean_absolute_error(y, base_results.predict(sm.add_constant(X_base)))
base_mae

In [ ]:
# Visualize model fit
fig, ax = plt.subplots(figsize=(15,5))
df.plot(x='sqft_living', y='price', kind="scatter", label="Data points", ax=ax)
sm.graphics.abline_plot(model_results=base_results, label="Regression line (radio)", c="red", linewidth=2, ax=ax)
ax.legend()
plt.savefig('images/model fit');
plt.show()

#Visualize fitted values
fig, ax = plt.subplots(figsize=(15,5))
sm.graphics.plot_fit(base_results, "sqft_living", ax=ax)
plt.savefig('images/fitted values');
plt.show()


#### Observations
Our model is statistically significant overall with a Prob (F-statistic) value of 0.00 which is less than the standard alpha 0.05. The model also explains about 29% of the variance in price. A higher variance is always preferred.

Both our intercept (const) and our coefficient for sqft_living are statistically significant with P>|t| values of 0.000 which are less than the standard alpha of 0.05.

Our intercept is about 144200, meaning that a home with 0 square feet of living area would cost about 144200$.

Our coefficient for sqft_living is about 169, which means that for each additional square foot of living area, we expect the price to increase by about 169$.

A mean absolute error (MAE) of 130267.37 means that our model is off by 130267$. This means that the difference between the actual price and the predicted price is approximately 130267$ which is quite high. A lower MAE is always preferred.

### 5.3 Multiple Regression Modeling

####  Setting Up Variables for Regression

In [ ]:
# variable representing all the highly correlated variables
X_multiple1 = df[["sqft_living", "grade", "sqft_living15", "sqft_above", "bathrooms"]]
X_multiple1= X_multiple1.astype(object)

#### Preprocessing Categorical Variable by One Hot-Encoding

The grade column represents the overall grade given to the housing unit, based on King County grading system. In order to use this variable in a model, it needs to be transformed and this will be done by creating multiple dummy variables, one for each category of the categorical variable.

In [ ]:
# one-hot encode the categorical column in X_iterated
X_multiple1 = pd.get_dummies(X_multiple1, columns=['grade'])

X_multiple1

#### Build a Multiple Linear Regression Model

Using the y variable from our base model and X_multiple1, we build a model called multiple1_model and a regression results object called multiple1_results.

In [ ]:
multiple1_model = sm.OLS(y, sm.add_constant(X_multiple1))
multiple1_results = multiple1_model.fit()

In [ ]:
#### Evaluate and Interpret Multiple Linear Regression Model Results
print(multiple1_results.summary())